Makefile

```diff
@@ -762,24 +769,43 @@ runtime/ld.conf: $(ROOTDIR)/Makefile.config
 # see http://pubs.opengroup.org/onlinepubs/9699919799/utilities/sort.html:
 # "using sort to process pathnames, it is recommended that LC_ALL .. set to C"
 
-# To speed up builds, we avoid changing "primitives" when files
-# containing primitives change but the primitives table does not
-runtime/primitives: \
-  $(shell runtime/gen_primitives.sh > runtime/primitives.new; \
-                    cmp -s runtime/primitives runtime/primitives.new || \
-                    echo runtime/primitives.new)
-	cp $^ $@
+# To speed up builds, we avoid changing "primitives" and
+# "primitive-arguments" when files containing primitives change but
+# the primitives table does not
+runtime/primitives-arguments.new: \
+  $(runtime_BUILT_HEADERS) \
+  $(runtime_PRIMITIVES) \
+  runtime/gen_primitives.sed runtime/gen_primitives_ints.sed
+	(for prim in $(runtime_PRIMITIVES); do \
+	     $(CPP) -I./runtime -D'CAMLprim()=0' "$${prim}" | \
+	     sed -n -f runtime/gen_primitives.sed; \
+	 done; \
+	 sed -n -f runtime/gen_primitives_ints.sed runtime/ints.c) | \
+	(export LC_ALL=C; sort | uniq) > $@
+
+runtime/primitives-arguments runtime/primitives: \
+  runtime/primitives-arguments.new
+	if ! cmp -s runtime/primitives-arguments $<; then \
+	    cp $< runtime/primitives-arguments; \
+	    sed -n -e 's/^\([a-z0-9_][a-z0-9_]*\).*$$/\1/p' \
+	      $< > runtime/primitives; \
+	fi
 
-runtime/prims.c : runtime/primitives
+runtime/prims.c : runtime/primitives-arguments
 	(echo '#define CAML_INTERNALS'; \
-         echo '#include "caml/mlvalues.h"'; \
+	 echo '#include "caml/mlvalues.h"'; \
 	 echo '#include "caml/prims.h"'; \
-	 sed -e 's/.*/extern value &();/' $<; \
+	 echo 'struct lexer_buffer; struct lexing_table;'; \
+	 echo 'struct parser_env; struct parser_tables;'; \
+	 echo; \
+	 sed -e 's/.*/extern value &;/' $<; \
+	 echo; \
 	 echo 'c_primitive caml_builtin_cprim[] = {'; \
-	 sed -e 's/.*/  &,/' $<; \
+	 sed -e 's/^\([a-z0-9_][a-z0-9_]*\) *(.*)/  (c_primitive)\1,/' $<; \
 	 echo '  0 };'; \
+	 echo; \
 	 echo 'char * caml_names_of_builtin_cprim[] = {'; \
-	 sed -e 's/.*/  "&",/' $<; \
+	 sed -e 's/^\([a-z0-9_][a-z0-9_]*\) *(.*)/  "\1",/' $<; \
```

Comment:
All this `sed` without setting `LC_CTYPE` makes me nervous.  Experience shows that it fails with bizarre locales.  Or, use named character classes, e.g. `[[:alnum]_]`.

Reply:
The comment above suggested `LC_ALL=C` instead of `LC_CTYPE`. I'm not familiar with these, which one should I use? In the meantime, I've using `export LC_ALL=C;` where needed.